ibis/mapd/operations.py

```diff
@@ -0,0 +1,644 @@
+    else:
+        raise NotImplementedError(type(expr))
+
+
+class CaseFormatter(object):
+
+    def __init__(self, translator, base, cases, results, default):
+        self.translator = translator
+        self.base = base
+        self.cases = cases
+        self.results = results
+        self.default = default
+
+        # HACK
+        self.indent = 2
+        self.multiline = len(cases) > 1
+        self.buf = StringIO()
+
+    def _trans(self, expr):
+        return self.translator.translate(expr)
+
+    def get_result(self):
+        """
+
+        :return:
+        """
+        self.buf.seek(0)
+
+        self.buf.write('CASE')
+        if self.base is not None:
+            base_str = self._trans(self.base)
+            self.buf.write(' {0}'.format(base_str))
+
+        for case, result in zip(self.cases, self.results):
+            self._next_case()
+            case_str = self._trans(case)
+            result_str = self._trans(result)
+            self.buf.write('WHEN {0} THEN {1}'.format(case_str, result_str))
+
+        if self.default is not None:
+            self._next_case()
+            default_str = self._trans(self.default)
+            self.buf.write('ELSE {0}'.format(default_str))
+
+        if self.multiline:
+            self.buf.write('\nEND')
+        else:
+            self.buf.write(' END')
+
+        return self.buf.getvalue()
+
+    def _next_case(self):
+        if self.multiline:
+            self.buf.write('\n{0}'.format(' ' * self.indent))
+        else:
+            self.buf.write(' ')
+
+
+def _table_array_view(translator, expr):
+    ctx = translator.context
+    table = expr.op().table
+    query = ctx.get_compiled_expr(table)
+    return '(\n{0}\n)'.format(util.indent(query, ctx.indent))
+
+
+def _timestamp_truncate(translator, expr):
+    op = expr.op()
+    arg, unit = op.args
+
+    if unit.upper() in [u.upper() for u in _mapd_unit_names.keys()]:
+        unit_ = _mapd_unit_names[unit].upper()
+    else:
+        raise ValueError('`{}` unit is not supported!'.format(unit))
+
+    # return _call_date_trunc(translator, converter, arg)
+    arg_ = translator.translate(arg)
+    return 'DATE_TRUNC({0!s}, {1!s})'.format(unit_, arg_)
+
+
+def _table_column(translator, expr):
+    op = expr.op()
+    field_name = op.name
+    quoted_name = quote_identifier(field_name, force=True)
+
+    table = op.table
+    ctx = translator.context
+
+    # If the column does not originate from the table set in the current SELECT
+    # context, we should format as a subquery
+    if translator.permit_subquery and ctx.is_foreign_expr(table):
+        proj_expr = table.projection([field_name]).to_array()
+        return _table_array_view(translator, proj_expr)
+
+    if ctx.need_aliases():
+        alias = ctx.get_ref(table)
+        if alias is not None:
+            quoted_name = '{}.{}'.format(alias, quoted_name)
+
+    return quoted_name
+
+
+# AGGREGATION
+
+class ApproxCountDistinct(ops.Reduction):
```

Comment:
This should be moved to `ibis.expr.api`

Reply:
`ApproxCountDistinct` is a alternative to` HLLCardinality` (https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L819) ... the difference is ApproxCountDistinct receive a `int` parameter.

Some info about this on MapD:

> APPROX_COUNT_DISTINCT(x, e) gives an approximate count of the value x, based on an expected error rate defined in e. The error rate is an integer value from 1 to 100. The lower the value of e, the higher the precision, and the higher the memory cost. Select a value for e based on the level of precision required. On large tables with large cardinalities, consider using APPROX_COUNT_DISTINCT when possible to preserve memory.

Reply:
I think mapD should compile `HLLCardinality` operation to `APPROX_COUNT_DISTINCT` with a default error rate for now. Later we can introduce a separate operation or the error argument to `HLLCardinality` itself, but that needs input argument validation for other backends because they don't support other than `error=None`.

@cpcloud What do You think?